src/psyclone/psyir/symbols/datasymbol.py

```diff
@@ -368,21 +371,39 @@ def constant_value(self, new_value):
                     "Error setting '{0}' constant value. A DataSymbol with a "
                     "constant value must be a scalar but a shape was found."
                     "".format(self.name))
-            try:
-                lookup = TYPE_MAP_TO_PYTHON[self.datatype]
-            except KeyError:
-                raise ValueError(
-                    "Error setting '{0}' constant value. Constant values are "
-                    "not supported for '{1}' datatypes."
-                    "".format(self.name, self.datatype))
-            if not isinstance(new_value, lookup):
-                raise ValueError(
-                    "Error setting '{0}' constant value. This DataSymbol "
-                    "instance datatype is '{1}' which means the constant "
-                    "value is expected to be '{2}' but found '{3}'."
-                    "".format(self.name, self.datatype,
-                              TYPE_MAP_TO_PYTHON[self.datatype],
-                              type(new_value)))
+
+            if isinstance(new_value, Node):
+                for node in new_value.walk(Node):
+                    if not isinstance(node, (Literal, Operation, Reference)):
+                        raise ValueError(
```

Comment:
Should this be a NotImplementedError? i.e. have we chosen to just support a sub-set of what is possible?

Reply:
Response in a comment below